pkg/controller/certrotation/certrotation_controller.go

```diff
@@ -0,0 +1,450 @@
+			Type: corev1.SecretTypeTLS,
+		}
+		// Recreate the secret. is this necessary? It will be updated in the first loop
+		if _, err := c.kubeClient.CoreV1().Secrets(ctrlcommon.MCONamespace).Create(context.Background(), newSecret, metav1.CreateOptions{}); err != nil {
+			return err
+		}
+		klog.Infof("Migration to %s for %s successful", corev1.SecretTypeTLS, ctrlcommon.MachineConfigServerTLSSecretName)
+	}
+
+	return nil
+}
+
+func (c *CertRotationController) addConfigMap(obj interface{}) {
+
+	configMap := obj.(*corev1.ConfigMap)
+
+	// Take no action if this isn't the machine-config-server-ca configMap
+	if configMap.Name != ctrlcommon.MachineConfigServerCAName {
+		return
+	}
+
+	klog.Infof("configMap %s added, reconciling all user data secrets", configMap.Name)
+
+	go func() {
+		c.reconcileUserDataSecrets()
+	}()
+}
+
+func (c *CertRotationController) updateConfigMap(oldCM, newCM interface{}) {
+
+	oldConfigMap := oldCM.(*corev1.ConfigMap)
+	newConfigMap := newCM.(*corev1.ConfigMap)
+
+	// Take no action if this isn't the machine-config-server-ca configMap
+	if oldConfigMap.Name != ctrlcommon.MachineConfigServerCAName {
+		return
+	}
+
+	// Only take action if the there is an actual change in the configMap
+	if oldConfigMap.ResourceVersion == newConfigMap.ResourceVersion {
+		return
+	}
+
+	klog.Infof("configMap %s updated, reconciling all user data secrets", oldConfigMap.Name)
+
+	// Reconcile all user data secrets
+	go func() {
+		c.reconcileUserDataSecrets()
+	}()
+}
+
+func (c *CertRotationController) addSecret(obj interface{}) {
+
+	secret := obj.(*corev1.Secret)
+
+	// Take no action if this is not a *-user-data secret
+	if !isUserDataSecret(*secret) {
+		return
+	}
+
+	klog.Infof("secret %s added, ensuring CA is up to date", secret.Name)
+
+	go func() { c.reconcileSecret(*secret) }()
+}
+
+func (c *CertRotationController) updateSecret(oldS, newS interface{}) {
+
+	oldSecret := oldS.(*corev1.Secret)
+	newSecret := newS.(*corev1.Secret)
+
+	// Take no action if this is not a *-user-data secret
+	if !isUserDataSecret(*oldSecret) {
+		return
+	}
+
+	// Only take action if the there is an actual change in the secret
+	if oldSecret.ResourceVersion == newSecret.ResourceVersion {
+		return
+	}
+
+	klog.Infof("secret %s updated, ensuring CA is up to date", newSecret.Name)
+
+	go func() { c.reconcileSecret(*newSecret) }()
+}
+
+func (c *CertRotationController) reconcileUserDataSecrets() {
+
+	mapiSecrets, err := c.maoSecretLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("cannot list MAO secrets: %v", err)
+		return
+	}
+	for _, secret := range mapiSecrets {
+		// Only reconcile *-user-data secrets
+		if !isUserDataSecret(*secret) {
+			continue
+		}
+		klog.V(4).Infof("Reconciling secret %s", secret.Name)
+		if err := c.reconcileSecret(*secret); err != nil {
+			klog.Errorf("error reconciling secret: %s %v", secret.Name, err)
+		}
+	}
+
+}
+
+func (c *CertRotationController) reconcileSecret(secret corev1.Secret) error {
```

Comment:
Just to double check, let's say something in this function errors, we'd probably log it in the controller, but it doesn't surface up any further than that right?

Reply:
Yep, it shouldn't cause any degrades, should be just a error log in this controller.